external/llvm-project/mlir/lib/ExecutionEngine/RocmRuntimeWrappers.cpp

```diff
@@ -202,7 +196,116 @@ mgpuMemGetDeviceMemRef1dInt32(int32_t *allocated, int32_t *aligned,
+                    int64_t stride2, int64_t stride3, int64_t stride4) {
+  float *gpuPtr;
+  HIP_REPORT_IF_ERROR(hipMalloc(
+      (void **)&gpuPtr, size0 * size1 * size2 * size3 * size4 * sizeof(float)));
+  return {gpuPtr,
+          gpuPtr,
+          offset,
+          {size0, size1, size2, size3, size4},
+          {stride0, stride1, stride2, stride3, stride4}};
+}
+
+extern "C" void mgpuMemDealloc5DFloat(
+    float *allocated, float *aligned, int64_t offset, int64_t size0,
+    int64_t size1, int64_t size2, int64_t size3, int64_t size4, int64_t stride0,
+    int64_t stride1, int64_t stride2, int64_t stride3, int64_t stride4) {
+  HIP_REPORT_IF_ERROR(hipFree(aligned));
+}
+
+extern "C" void mgpuMemCopy5DFloat(
+    float *sourceAllocated, float *sourceAligned, int64_t sourceOffset,
+    int64_t sourceSize0, int64_t sourceSize1, int64_t sourceSize2,
+    int64_t sourceSize3, int64_t sourceSize4, int64_t sourceStride0,
+    int64_t sourceStride1, int64_t sourceStride2, int64_t sourceStride3,
+    int64_t sourceStride4, float *destAllocated, float *destAligned,
+    int64_t destOffset, int64_t destSize0, int64_t destSize1, int64_t destSize2,
+    int64_t destSize3, int64_t destSize4, int64_t destStride0,
+    int64_t destStride1, int64_t destStride2, int64_t destStride3,
+    int64_t destStride4, unsigned copyDirection) {
+  HIP_REPORT_IF_ERROR(hipMemcpy(destAligned, sourceAligned,
+                                sourceSize0 * sourceSize1 * sourceSize2 *
+                                    sourceSize3 * sourceSize4 * sizeof(float),
+                                static_cast<hipMemcpyKind>(copyDirection)));
+}
+
+extern "C" StridedMemRefType<unsigned short, 5>
+mgpuMemAlloc5DHalf(unsigned short *allocated, unsigned short *aligned,
+                   int64_t offset, int64_t size0, int64_t size1, int64_t size2,
+                   int64_t size3, int64_t size4, int64_t stride0,
+                   int64_t stride1, int64_t stride2, int64_t stride3,
+                   int64_t stride4) {
+  unsigned short *gpuPtr;
+  HIP_REPORT_IF_ERROR(
+      hipMalloc((void **)&gpuPtr, size0 * size1 * size2 * size3 * size4 *
+                                      sizeof(unsigned short)));
+  return {gpuPtr,
+          gpuPtr,
+          offset,
+          {size0, size1, size2, size3, size4},
+          {stride0, stride1, stride2, stride3, stride4}};
+}
+
+extern "C" void mgpuMemDealloc5DHalf(unsigned short *allocated,
+                                     unsigned short *aligned, int64_t offset,
+                                     int64_t size0, int64_t size1,
+                                     int64_t size2, int64_t size3,
+                                     int64_t size4, int64_t stride0,
+                                     int64_t stride1, int64_t stride2,
+                                     int64_t stride3, int64_t stride4) {
+  HIP_REPORT_IF_ERROR(hipFree(aligned));
+}
+
+extern "C" void mgpuMemCopy5DHalf(
```

Comment:
Aside from the deallocator, can we investigate ways to not use these? If not here, then could you file a tiket so we get around to it later?

Reply:
@krzysz00 so why is it not good to use these?

Reply:
@krzysz00 I noticed that for some gpu runtime functions, their high level IR functions are used in mlir (like gpu.* functions). And for others ( which seem to be not basic), the runtime mgpu* functions are generated directly. Is there a way to determine when to use the basic functions (gpu.*) and when to use non-basic functions (mgpu*)?
Is this related to the new ticket [MLIR#555](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/555)?

Reply:
Yeah

The historical situation is that we called those runtime wrappers directly when generating tests (I wasn't around when we did that, and this was before we started having a split between `external/` and `mlir/`). Then, upstream, at some point, added `gpu.*` wrappers around these functions, and we didn't move to them.

The long-term plan is, I think, to move to these `gpu.*` wrappers to make things less fragile and to make our code clearer. However, there are some functions that don't have `gpu.*` equivalents yet, and for those we need to create them or find a way to not use them anymore. (ex. these 5D memcopy functions could probably be replaced by a wrapper that extracts the pointer from the memref and calls a 1D memcopy, keeping the C API small.

Reply:
Apologies, I already started this on https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/555.

I can pass off to you @zhanglx13 if you like. It is pretty simple change but there are alot of tests to change.

Reply:
@sjw36 Yes you can assign that ticket to me.

Reply:
@zhanglx13 Just a reminder that if you decide to do that one, implement that from a independent PR. I feel like there's already enough of the code changes in this one, and we want to regress it as its own feature.

Reply:
@jerryyin Sure. I'll merge this one as soon as I get all tests passing.